[contrib][linux] Fix a warning in zstd_reset_cstream()

cyberknight777 opened the pull request:
- This fixes the below warning:

../lib/zstd/zstd_compress_module.c: In function 'zstd_reset_cstream':
../lib/zstd/zstd_compress_module.c:136:9: warning: 'ZSTD_resetCStream' is deprecated [-Wdeprecated-declarations]
  136 |         return ZSTD_resetCStream(cstream, pledged_src_size);
      |         ^~~~~~
In file included from ../include/linux/zstd.h:26,
                 from ../lib/zstd/zstd_compress_module.c:15:
../include/linux/zstd_lib.h:2277:8: note: declared here
 2277 | size_t ZSTD_resetCStream(ZSTD_CStream* zcs, unsigned long long pledgedSrcSize);
      |        ^~~~~~~~~~~~~~~~~

ZSTD_resetCstream is deprecated and zstd_CCtx_reset is suggested to use hence let's switch to it.

Signed-off-by: Cyber Knight <cyberknight755@gmail.com>

cyberknight777 committed:
[contrib][linux] Fix a warning in zstd_reset_cstream()

- This fixes the below warning:

../lib/zstd/zstd_compress_module.c: In function 'zstd_reset_cstream':
../lib/zstd/zstd_compress_module.c:136:9: warning: 'ZSTD_resetCStream' is deprecated [-Wdeprecated-declarations]
  136 |         return ZSTD_resetCStream(cstream, pledged_src_size);
      |         ^~~~~~
In file included from ../include/linux/zstd.h:26,
                 from ../lib/zstd/zstd_compress_module.c:15:
../include/linux/zstd_lib.h:2277:8: note: declared here
 2277 | size_t ZSTD_resetCStream(ZSTD_CStream* zcs, unsigned long long pledgedSrcSize);
      |        ^~~~~~~~~~~~~~~~~

ZSTD_resetCstream is deprecated and zstd_CCtx_reset is suggested to use hence let's switch to it.

Signed-off-by: Cyber Knight <cyberknight755@gmail.com>

Cyan4973 commented:
This is kernel code, so there might be specific reasons.
cc @terrelln .

cyberknight777 commented:
Hmm I've never gotten this warn when compiling ZSTD 1.5.2 with the Linux kernel. Only warn I've got was the one which I fixed and made the PR with. I'll try another way to fix the warn completely without triggering any other warnings.

terrelln commented on contrib/linux-kernel/zstd_compress_module.c:
```diff
@@ -133,7 +133,7 @@ EXPORT_SYMBOL(zstd_init_cstream);
 	unsigned long long pledged_src_size)
 {
-	return ZSTD_resetCStream(cstream, pledged_src_size);
+	return ZSTD_CCtx_reset(cstream, pledged_src_size);
```
This is incorrect, this is the definition of `ZSTD_resetCStream()`. You should do the same here.

https://github.com/facebook/zstd/blob/930b5cecaa033c3aa6017b246bb804ef576942b4/lib/compress/zstd_compress.c#L5204-L5212

E.g. 

```c
if (pledged_src_size == 0)
    pledged_src_size = ZSTD_CONTENTSIZE_UNKNOWN;
ZSTD_FORWARD_IF_ERR( ZSTD_CCtx_reset(zstream, ZSTD_reset_session_only) ); 
ZSTD_FORWARD_IF_ERR( ZSTD_CCtx_setPledgedSrcSize(zstream, pledged_src_size) );
return 0;
```

cyberknight777 replied:
How about changing cstream to zcs and pledged_src_size to pledgedSrcSize in contrib/linux-kernel/zstd_compress_module.c? This would be a better implementation than to modify the core function of ZSTD_CCtx_reset which might break the functionality of it in other use cases.

cyberknight777 replied:
Update: I just tested with ZSTD_CCtx_setPledgedSrcSize and it seemed to work fine(except I had to change pledged_src_size to pledgedSrcSize in zstd_compress_module.c).

I'll push a patch with the change I stated.

cc: @terrelln

terrelln replied:
@cyberknight777 I'm not asking you to change `ZSTD_CCtx_reset()`, I'm asking you to make `zstd_reset_cstream()` functionally identical to `ZSTD_resetCStream()`.

You still need to call `ZSTD_CCtx_reset()`. Also, please keep the name `pledged_src_size`.

The function should be (with tab indentation):

```c
size_t zstd_reset_cstream(zstd_cstream *cstream,
	unsigned long long pledged_src_size)
{
    if (pledged_src_size == 0)
        pledged_src_size = ZSTD_CONTENTSIZE_UNKNOWN;
    ZSTD_FORWARD_IF_ERR( ZSTD_CCtx_reset(cstream, ZSTD_reset_session_only) ); 
    ZSTD_FORWARD_IF_ERR( ZSTD_CCtx_setPledgedSrcSize(cstream, pledged_src_size) );
    return 0;
}
```

cyberknight777 replied:
Alright, I see what you mean. I'm pushing another patch in a few.

cyberknight777 replied:
@terrelln any comments on the patch I pushed?

cyberknight777 committed:
[contrib][linux] Use ZSTD_CCtx_setPledgedSrcSize() instead of ZSTD_CCtx_reset()

- The previous patch throws the following warning:

 ../linux/lib/zstd/zstd_compress_module.c: In function ‘zstd_reset_cstream’:
../linux/lib/zstd/zstd_compress_module.c:136:34: error: enum conversion when passing argument 2 of ‘ZSTD_CCtx_reset’ is invalid in C++ [-Werror=c++-compat]
  136 |  return ZSTD_CCtx_reset(cstream, pledged_src_size);
      |                                  ^~~~~~~~~~~~~~~~
In file included from ../linux/include/linux/zstd.h:26,
                 from ../linux/lib/zstd/zstd_compress_module.c:15:
../linux/include/linux/zstd_lib.h:501:20: note: expected ‘ZSTD_ResetDirective’ {aka ‘enum <anonymous>’} but argument is of type ‘long long unsigned int’
  501 | ZSTDLIB_API size_t ZSTD_CCtx_reset(ZSTD_CCtx* cctx, ZSTD_ResetDirective reset);
      |                    ^~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

Since we have a choice to either use ZSTD_CCtx_reset or ZSTD_CCtx_setPledgedSrcSize instead of ZSTD_resetCStream, let's switch to ZSTD_CCtx_setPledgedSrcSize to not have any unnecessary warns alongside the kernel build and CI test build.

Signed-off-by: Cyber Knight <cyberknight755@gmail.com>

cyberknight777 committed:
[contrib][linux] Make zstd_reset_cstream() functionally identical to ZSTD_resetCStream()

- As referenced by Nick Terrelln ~ the ZSTD maintainer in the linux kernel, making zstd_reset_cstream() functionally identical to ZSTD_resetCStream() would be the perfect way to fix the warning without touching any core functions or breaking other parts of the code.

Suggested-by: Nick Terrell <terrelln@fb.com>
Signed-off-by: Cyber Knight <cyberknight755@gmail.com>

terrelln reviewed (approved):
Thanks @cyberknight777!
